Module 2 HW

leg0ushque opened the pull request:
Counter, SearchForm, GenreSelect components

leg0ushque committed:
Create an app with create-react-app

leg0ushque committed:
Clear CRA

leg0ushque committed:
Create Counter component

leg0ushque committed:
Fix Counter import after folder renaming

leg0ushque committed:
Add mockup font

leg0ushque committed:
Refactoring

Remove redundant comments, put ctors before methods

leg0ushque committed:
Create SearchForm component

leg0ushque committed:
Update SearchForm component & styles

leg0ushque committed:
Create GenreSelect component

leg0ushque committed:
Update App.js to use all the components

leg0ushque committed:
Merge pull request #1 from leg0ushque/feature/module2-homework

Module 2 homework

anastasiyaBH commented on src/App.css:
```diff
@@ -0,0 +1,13 @@
+  font-family: "Montserrat";
+  src: local("Montserrat"),
+    url("../public/fonts/Montserrat/Montserrat-Regular.ttf") format("truetype");
+}
```
Please, add an empty line at the end of the file

leg0ushque replied:
Done (1b556c5a2b5407b219cf4e1953e78c67400cb18f)

anastasiyaBH commented on src/components/Counter/Counter.js:
```diff
@@ -0,0 +1,57 @@
+
+  constructor(props) {
+    super(props);
+    this.state.value = props.value;
```
Let's rename props.value to props.**initialValue**. 

https://ru.reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

leg0ushque replied:
Done (46f3b898fc701ef3a4d0fd3315d20036399fe336)

anastasiyaBH commented on src/components/Counter/index.js:
```diff
@@ -0,0 +1,3 @@
+import Counter from "./Counter";
+
+export { Counter };
```
We can also write this as
`
export default Counter
`

leg0ushque replied:
Fixed for all components (b0cb3c88d572bc6a2671a7f5afb18272e51cc346)

anastasiyaBH commented on src/components/Counter/styles.css:
```diff
@@ -0,0 +1,21 @@
+
+.decButton {
+  background-color: indianred;
+}
```
emplty line

leg0ushque replied:
Done (1b556c5a2b5407b219cf4e1953e78c67400cb18f)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+import React, { Component } from "react";
+import "./styles.css";
+
+const defaultValue = "";
```
Please, use UPPER_CASE for constants

leg0ushque replied:
Done for SearchForm (46f3b898fc701ef3a4d0fd3315d20036399fe336)
Fixed for other components (acb780a178648469d974bc88446542caddf0576e)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+
+  constructor(props) {
+    super(props);
+    this.state.defaultValue = props.defaultValue || defaultValue;
```
props.defaultValue -> props.initialValue

leg0ushque replied:
Fixed (46f3b898fc701ef3a4d0fd3315d20036399fe336)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  search = () => {
+    let searchInput = document.getElementById("searchInput");
```
Let's not use getting elements from DOM. It's better to use component state in React.

Please, review "Controlled and uncontrolled components", "DOM vs Virtual DOM".
https://ru.reactjs.org/docs/uncontrolled-components.html
https://goshacmd.com/controlled-vs-uncontrolled-inputs-react/
https://ru.reactjs.org/docs/faq-internals.html#gatsby-focus-wrapper

leg0ushque replied:
Made SearchForm a controlled component (199e1d5e57dd8e523ce5eca19af3f65fc5d22fc3)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+            placeholder="What do you want to watch?"
+            id="searchInput"
+            className="search-input"
+            defaultValue={this.state.defaultValue}
```
Let's create controlled component. A good example you cand find here.
https://goshacmd.com/controlled-vs-uncontrolled-inputs-react/

leg0ushque replied:
Made SearchForm a controlled component (199e1d5e57dd8e523ce5eca19af3f65fc5d22fc3)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+
+  search = () => {
+    let searchInput = document.getElementById("searchInput");
+    this.state.onSearch(searchInput.value);
```
It would be great to clear the value of the input after submitting the search.

leg0ushque replied:
Made SearchForm a controlled component, the input is cleared after the form submitted (199e1d5e57dd8e523ce5eca19af3f65fc5d22fc3)
Empty value search is not allowed (adada93ae61db74c07c0909cf2cc5150155eacc3)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+          <input
+            type="text"
+            placeholder="What do you want to watch?"
+            id="searchInput"
```
Please don't use DOM id

leg0ushque replied:
Fixed. Made SearchForm a controlled component(199e1d5e57dd8e523ce5eca19af3f65fc5d22fc3)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+  render() {
+    return (
+      <>
+        <div className="search-form">
```
Your search callback doesn't call when **the user presses Enter when the input has focus**. This is a significant part of the task. Review this topic for help https://ru.reactjs.org/docs/forms.html#controlled-components

leg0ushque replied:
Made SearchForm a controlled component. Form is submitted on having focus on input and pressing Enter
(199e1d5e57dd8e523ce5eca19af3f65fc5d22fc3)

anastasiyaBH commented on src/components/SearchForm/SearchForm.jsx:
```diff
@@ -0,0 +1,43 @@
+
+  render() {
+    return (
+      <>
```
You can remove Fragments here. https://ru.reactjs.org/docs/fragments.html

leg0ushque replied:
Fixed in every component (d1751c51277a946462dc0a1948c8bc82d9ee253a)

anastasiyaBH commented on src/components/SearchForm/styles.css:
```diff
@@ -0,0 +1,54 @@
+
+  text-align: center;
+  text-transform: uppercase;
+}
```
empty line

leg0ushque replied:
Done (1b556c5a2b5407b219cf4e1953e78c67400cb18f)

anastasiyaBH commented on src/components/GenreSelect/GenreSelect.jsx:
```diff
@@ -0,0 +1,56 @@
+    return (
+      <li
+        className={selected ? "genreButton selected" : "genreButton"}
+        genre={genreName}
```
Let's don't use attributes.

leg0ushque replied:
Fixed, _genre_ attribute removed

anastasiyaBH commented on src/components/GenreSelect/GenreSelect.jsx:
```diff
@@ -0,0 +1,56 @@
+        className={selected ? "genreButton selected" : "genreButton"}
+        genre={genreName}
+        onClick={(e) =>
+          !selected ? this.select(e.target.attributes.genre.value) : {}
```
You can write here as `this.select(genreName)`

leg0ushque replied:
Fixed (219ec1b7ac65f09472d621398729aebe17a73e6f)

anastasiyaBH commented on src/components/GenreSelect/GenreSelect.jsx:
```diff
@@ -0,0 +1,56 @@
+  createGenreButton(genreName, selected) {
+    return (
+      <li
+        className={selected ? "genreButton selected" : "genreButton"}
```
Add props key here. Why is it handy? https://ru.reactjs.org/docs/lists-and-keys.html

leg0ushque replied:
Fixed

anastasiyaBH commented on src/components/GenreSelect/GenreSelect.jsx:
```diff
@@ -0,0 +1,56 @@
+
+  constructor(props) {
+    super(props);
+    this.state.genres = Array.from(props.genres) || new Array() < "" > 0;
```
Why do you write this? 
`Array.from(props.genres) || new Array() < "" > 0;`

leg0ushque replied:
It didn't accept _this.state.genres_ as Array, so map(...) was imposible. The right part of the expression was supposed to be new Array()<"">(0) with zero capacity, but something went wrong 😂. 
Simplified with "[]" (acb780a178648469d974bc88446542caddf0576e)

anastasiyaBH commented on src/components/GenreSelect/GenreSelect.jsx:
```diff
@@ -0,0 +1,56 @@
+    );
+  }
+
+  createGenreButton(genreName, selected) {
```
Let's move this function before render()

leg0ushque replied:
Fixed (90f105036af1f13bbc71f4054a923bdfee5320e2)

anastasiyaBH commented on src/components/GenreSelect/GenreSelect.jsx:
```diff
@@ -0,0 +1,56 @@
+    this.state.onSelect(genreName);
+  };
+
+  render() {
```
When you click on a specific genre, the section is not highlighted. I guess you need to store the state of the selected genre in the parent component and change it on onSelect.

https://www.pluralsight.com/guides/how-to-handle-communication-between-parent-and-child-components-in-reactjs

leg0ushque replied:
Fixed (61a6a37713a95381789739439d3619d433b7c6ba)

anastasiyaBH reviewed (commented):
Good job. Please, review my notes regarding the code and make some changes

leg0ushque committed:
Add an empty line at the end of every .css file.

leg0ushque committed:
Props now provide 'initialValue' instead of 'value'.

leg0ushque committed:
Use 'default' in index.js of every component.

leg0ushque committed:
Use SNAKE_UPPERCASE notation for constants.

Update GenresButtons approach to add 'All' button using spread operator.
Replace 'new Array()<"">(0) with '[]'

leg0ushque committed:
Use parameter instead of using event targer attributes

Prettier styles fixes included

leg0ushque committed:
Move createGenreButton function before render()

leg0ushque committed:
Remove redundant Fragments

leg0ushque committed:
Update SearchForm component to be controlled

leg0ushque committed:
Made SearchField input a required one to prevent empty value search

leg0ushque committed:
Attempt to fix GenreSelect

leg0ushque committed:
Fixed GenreSelect component
